File: Algo/Leetcode/503NextGreaterNumII.cpp
// dict + 单调栈
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        stack<int> s;

        vector<int> res(n, -1);
        for(int i=0; i<nums.size()*2-1; i++){
            while(!s.empty() && nums[s.top() % n] < nums[i % n]){
                res[s.top() % n] = nums[i % n];
                s.pop();
            }
            s.push(i);
        }
        return res;
    }
};

File: Algo/Leetcode/503NextGreaterNumII.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        _nums = nums + nums[:-1]
        n = len(nums)
        res = [-1] * n

        st = []
        for i, k in enumerate(_nums):
            while len(st) > 0 and st[-1][1] < k:
                top = st.pop()
                res[top[0] % n] = k
            st.append((i, k))

        return res